### PR TITLE
Editable install: do not show any output in  verbose mode when there is no work to do

### PR DESCRIPTION
**Pull request by lesteve**
### Why the current behaviour is problematic

The "ninja: no work to do" output can be confusing as noted for example by @ogrisel in https://github.com/scikit-learn/scikit-learn/pull/28040#issuecomment-1886676007. @eli-schwartz mentioned that it could potentially be changed in meson-python https://github.com/scikit-learn/scikit-learn/pull/28040#issuecomment-1888470251.

Personally, when working on scikit-learn code, I find it reassuring to see some output if the sklearn import takes a while and get the feed-back that some compilation is happening. In the case when there is nothing to do I would rather have nothing printed. One thing I realised recently is that the "no work to do" output can be printed multiple times if you use subprocesses (makes complete sense since sklearn is imported in each subprocess). This can happen easily in scikit-learn for example with this snippet.

```py
# /tmp/test.py
from sklearn.linear_model import LogisticRegression
from sklearn.datasets import make_classification
from sklearn.model_selection import cross_val_score

model = LogisticRegression()
X, y = make_classification()

print(cross_val_score(model, X, y, n_jobs=4))
```

Output is:
```
❯ python /tmp/test.py
+ /home/lesteve/micromamba/envs/scikit-learn-dev/bin/ninja
ninja: no work to do.
+ /home/lesteve/micromamba/envs/scikit-learn-dev/bin/ninja
+ /home/lesteve/micromamba/envs/scikit-learn-dev/bin/ninja
+ /home/lesteve/micromamba/envs/scikit-learn-dev/bin/ninja
+ /home/lesteve/micromamba/envs/scikit-learn-dev/bin/ninja
ninja: no work to do.
ninja: no work to do.
ninja: no work to do.
ninja: no work to do.
[1.   0.85 0.9  0.95 0.95]
```

### Current implementation

The simplest thing I could find was to first execute the `ninja` command in dry-run mode and to check if "no work to do" is in the captured stdout. This may be a bit brittle in case the ninja output changes. Maybe there is a better way to check with `ninja` that there is no work to do?

There does not seem to be any test right now for the editable verbose behavior, but I can try to add some if you think this is worth it.

### Alternative implementation

If you think the overhead of having a `ninja` dry-run command is too much overhead, one possible alternative implementation would be to capture `stdout` with `p = subprocess.Popen(..., stdout=subprocess.PIPE)` and look at the first two lines of output. In my experience, this kind of manipulation are always more complicated than you would hope, but maybe I have been doing it wrong all this time ...

The kind of complications I have in mind:
- `p.stdout.read` is blocking so in the case that there is no output for some reason, it will wait for ever. Maybe we know there will always be some output with the `ninja` command. You can also put you can use `fcntl` to put stdout in non-blocking mode.
- displaying the captured stdout live (i.e. without waiting for the command to finish) is also always more complicated than you would hope:
  + decide whether to read line by line, a few lines, a number of bytes
  + decide whether to sleep between two reads (I have seen some cases where not sleeping had a high CPU usage)
  + you also need to remember to redisplay all the remaining stdout once the command finishes otherwise you miss some stdout.

### CODE REVIEW

**Comment by ogrisel** [comment]
I think the lines:

```
+ /home/lesteve/micromamba/envs/scikit-learn-dev/bin/ninja
ninja: no work to do.
```

would be ok, if they were prefixed by some user friendlier message:

```
meson-python: importing sklearn installed in editable mode
+ /home/lesteve/micromamba/envs/scikit-learn-dev/bin/ninja
ninja: no work to do.
```

But I agree that the repeated message is a bit annoying when using parallelism that spawn Python workers that import `sklearn` concurrently.

**Comment by ogrisel** [comment]
But I am also fine with what @lesteve suggests in this PR.

File: mesonpy/_editable.py
Comment (eli-schwartz) [inline mesonpy/_editable.py:310]: Relevant: https://github.com/mesonbuild/meson/blob/a35d4d368a21f4b70afa3195da4d6292a649cb4c/mesonbuild/mtest.py#L1636

**Comment by rgommers** [comment]
Thanks! I agree that this message can be confusing. I hadn't seen it printed multiple times yet, but that makes it worse. So looks like a good idea to suppress it.

File: mesonpy/_editable.py
Comment (rgommers) [reply mesonpy/_editable.py:310]: already tested code, always nice:)

Comment (rgommers) [reply mesonpy/_editable.py:310]: Probably worth taking that over verbatim, and adding a code comment showing its origin so that if we need to make further tweaks, we can keep the two pieces of code in sync?

Comment (eli-schwartz) [reply mesonpy/_editable.py:310]: That would be my recommendation, yes.

File: mesonpy/_editable.py
Comment (ogrisel) [inline mesonpy/_editable.py:312]: I find the build command not very informative because it does not reflect which editable install is being rebuilt, nor does it state that `meson-python`'s editable install is in charge of making this build happen.

So maybe consider:

```suggestion
                module_names = " ".join(sorted(self._top_level_modules))
                print(f"meson-python: building {module_names}")
```

Here is the resulting user experience:

```
$ python -c "import sklearn"  # nothing to rebuild => no output
$ touch sklearn/ensemble/_gradient_boosting.pyx
$ python -c "import sklearn"
meson-python: building sklearn
[3/3] Linking target sklearn/ensemble/_gradient_boosting.cpython-311-darwin.so
```

Comment (rgommers) [reply mesonpy/_editable.py:312]: I like it!

Comment (lesteve) [reply mesonpy/_editable.py:312]: Two questions:
- do we want to keep the `ninja` build command for troubleshooting or are we fine not showing it?
- do we want to keep the `+` at the beginning of the first line? I guess maybe this was a marker to make it stand out in the output?

File: mesonpy/_editable.py
Comment (dnicolodi) [inline mesonpy/_editable.py:309]: This does not work on Windows, where the compilation command is `meson compile`, which does not accept a `-n` argument. This demonstrates that this feature needs tests.

**Review by dnicolodi — CHANGES_REQUESTED** [review]

File: mesonpy/_editable.py
Comment (ogrisel) [reply mesonpy/_editable.py]: I don't have an opinion on those points as long as the first message mentions "meson-python" and the name of the package(s) being rebuilt prior to displaying build outputs.

File: tests/test_editable.py
Comment (dnicolodi) [inline tests/test_editable.py:219]: Why do you check for the exact content of the log messages? It seems fragile (they are not under the control of meson-python) and not really required. The output is also very different on Windows.

File: tests/test_editable.py
Comment (dnicolodi) [inline tests/test_editable.py:215]: The compilation command on Windows is neither `ninja` nor `samu`.

File: mesonpy/_editable.py
Comment (dnicolodi) [inline mesonpy/_editable.py:316]: For consistency, I would spell this `meson-python: executing {build_command}`.

Comment (dnicolodi) [reply mesonpy/_editable.py:316]: Or you can make it a single line, combining the two messages: `meson-python: building {module_names}: {build_command}`. And I would loose the `+ ` prefix.

File: mesonpy/_editable.py
Comment (dnicolodi) [inline mesonpy/_editable.py:310]: Right now, this is failing on Windows, where `meson compile -n` cannot work.

File: tests/test_editable.py
Comment (lesteve) [reply tests/test_editable.py:219]: I was a bit overoptimistic indeed, I am going to only check the first two lines of output since they are under meson-python control.

**Comment by lesteve** [comment]
It would be great if someone with the necessary rights could enable the Github actions workflow :pray:! 

I can reproduce the Alpine failure in a Docker container and the bad news is that `ninja -n` seems to segfault (`ninja` alone works fine). The ninja version is 1.9 (released January 2019) and I have seen at least one test in meson-python that is skipped if the ninja version < 1.10, so maybe this could be an option?

My basic gdb skills only allowed me to get this information:
```
(gdb) r
Starting program: /usr/bin/ninja -C build/cp312 -n
warning: Error disabling address space randomization: Operation not permitted
ninja: entering directory 'build/cp312'

Program received signal SIGSEGV, Segmentation fault.
0x00007b65464a9458 in strlen (s=<optimized out>, 
    s@entry=0x8 <error: Cannot access memory at address 0x8>) at src/string/strlen.c:17
warning: 17	src/string/strlen.c: No such file or directory
```

Suggestions how to get more info more than welcome ...

**Comment by eli-schwartz** [comment]
> I can reproduce the Alpine failure in a Docker container and the bad news is that `ninja -n` seems to segfault (`ninja` alone works fine). The ninja version is 1.9 (released January 2019) and I have seen at least one test in meson-python that is skipped if the ninja version < 1.10, so maybe this could be an option?

This is:

https://github.com/michaelforney/samurai/issues/66
https://github.com/michaelforney/samurai/issues/81
https://github.com/michaelforney/samurai/commit/fb61f22c7e690715d309c41812412c4f432ef53a

Possibly something Alpine should backport...

**Comment by lesteve** [comment]
I have skipped the test on Alpine for now based on detecting musllinux.

To know whether the test pass on Windows it would be nice that someone enable Github workflows. Full disclosure: locally the Windows tests don't pass and I don't really know why yet.

**Comment by rgommers** [comment]
> To know whether the test pass on Windows it would be nice that someone enable Github workflows.

Done. this can be a pain because the approve button has to be pressed after every push. If you want to submit a trivial PR to side-step this problem, I'm happy to merge that straight away (e.g, in the readme, change `to use Meson` to `using Meson`).

**Comment by lesteve** [comment]
Thanks, the Windows tests fail in a similar way locally, the dry-run command fails with a exit code 1 and I can't find a reason in the captured stderr as to why, but maybe I am missing something ...

I'll be offline next week but I can try to pick this up when I come back. Any meson-python maintainer should not hesitate to push into my branch if that makes things easier.

File: mesonpy/_editable.py
Comment (dnicolodi) [inline mesonpy/_editable.py:322]: Assigning the message to a variable is unnecessary.

File: mesonpy/_editable.py
Comment (dnicolodi) [inline mesonpy/_editable.py:309]: The build command on Windows can be something else that the literal string `meson`.

File: tests/test_editable.py
Comment (dnicolodi) [inline tests/test_editable.py:223]: The past 10 lines can simply be:
```
assert output.startswith('meson-python: building complex with')
```

File: tests/test_editable.py
Comment (lesteve) [reply tests/test_editable.py:210]: As the comment above is trying to say, touching the `.pyx` ensures that recompilation happens.

File: mesonpy/_editable.py
Comment (lesteve) [reply mesonpy/_editable.py]: Good catch the quoting was the issue indeed, now the test passes locally on a Windows VM.

About multiple `--ninja-args`, you probably know a lot more than me about this, so any suggestions welcome!

File: mesonpy/_editable.py
Comment (lesteve) [reply mesonpy/_editable.py:322]: Do you have any advice on a more robust way to handle the Windows case?

**Comment by lesteve** [comment]
> If you want to submit a trivial PR to side-step this problem, I'm happy to merge that straight away (e.g, in the readme, change `to use Meson` to `using Meson`).

@rgommers I have opened https://github.com/mesonbuild/meson-python/pull/585 with this change.

**Comment by rgommers** [comment]
> @rgommers I have opened #585 with this change.

Sounds good - merged now. Should fix the issue with CI running here.

File: mesonpy/_editable.py
Comment (dnicolodi) [reply mesonpy/_editable.py]: I don't know whether `meson compile` accepts multiple `--ninja-args` options. This is why I am asking. If I knew I would not have posed this as a question. You are the proponent of this feature, thus it is on you to verify and to find a solution.

File: tests/test_editable.py
Comment (dnicolodi) [reply tests/test_editable.py:213]: The comment above is precisely the reason why I am asking. "to touch" a file is usually understood as in updating the mtime in the file metadata without altering the file content. What you are doing here is to rewrite the content of the file. Which is, per your comment, unnecessary.

**Comment by dnicolodi** [comment]
> This is:
> 
> [michaelforney/samurai#66](https://github.com/michaelforney/samurai/issues/66) [michaelforney/samurai#81](https://github.com/michaelforney/samurai/issues/81) [michaelforney/samurai@fb61f22](https://github.com/michaelforney/samurai/commit/fb61f22c7e690715d309c41812412c4f432ef53a)
> 
> Possibly something Alpine should backport...

@eli-schwartz Are you implying that Alpine uses samurai but calls it ninja? I wonder if we can pull in a fixed package in our CI image...

**Comment by dnicolodi** [comment]
Yes, it seems that installing ninja on Alpine Linux puts the executable in a location outside `$PATH` and that the samuari package install a symlink that makes ninja actually be samurai. samurai seems to be installed with the build-base package, otherwise I don't understand why we get it in our CI images. Anyhow, installing `ninja-is-really-ninja` should make the `ninja` command actually be what is expected.

The bug @eli-schwartz mentions was fixed in samurai back in 2021, but samurai hasn't seen a release since 2020.

**Comment by lesteve** [comment]
The CI failures for the development meson version are also `main` and do not seem to be related to this PR.

File: tests/test_editable.py
Comment (lesteve) [reply tests/test_editable.py]: I changed this to only `touch` the file

File: mesonpy/_editable.py
Comment (dnicolodi) [reply mesonpy/_editable.py]: It is not acceptable to discard the user supplied argument and keep only the `-n`.

File: mesonpy/_editable.py
Comment (dnicolodi) [reply mesonpy/_editable.py:322]: The easiest is probably to check for `if sys.platform == 'win32'`. If that is true, the command being executed is `$MESON compile` possibly with a `--ninja-args=...` argument. Otherwise is `ninja` or something with a compatible command line interface.

File: mesonpy/_editable.py
Comment (dnicolodi) [inline mesonpy/_editable.py:334]: Why does Codecov think that these lines are not covered by the tests?

Comment (lesteve) [reply mesonpy/_editable.py:334]: Hmmm not sure what is happening ...

Comment (lesteve) [reply mesonpy/_editable.py:336]: It seems like a but somewhere in the toolchain (maybe `coverage`, maybe `codecov`, not sure ...) that does not see the line covered although the line is executed. 

- codecov says the line is not covered [build log](https://github.com/mesonbuild/meson-python/pull/579/checks?check_run_id=22255522195) or
image

- the line is executed as you can see in this [build log](https://github.com/mesonbuild/meson-python/actions/runs/8143472943/job/22255402547?pr=579#step:9:204) for example

File: mesonpy/_editable.py
Comment (lesteve) [reply mesonpy/_editable.py]: Point taken, my current plan for this is to fine the last arguments in `self._build_cmd` that starts with `--ninja-args` and if there is one to append `,-n` to it. This does seem a bit complicated though, let me know if you have a better suggestion!

In order to test this (at least locally), do you have an example in mind on Windows where ninja arguments would be already in the build command (i.e. before adding `-n`)?

Comment (eli-schwartz) [reply mesonpy/_editable.py]: My assumption is this would happen when users specify `-C...` options to the build frontend.

File: mesonpy/_editable.py
Comment (ogrisel) [reply mesonpy/_editable.py:336]: Maybe sklearn's editable install `find_spec` is called prior to the initialization of the `coverage` tracer?

Comment (dnicolodi) [reply mesonpy/_editable.py:336]: > Maybe sklearn's editable install's

What does sklearn has to do with this?

File: mesonpy/_editable.py
Comment (dnicolodi) [reply mesonpy/_editable.py]: As @eli-schwartz says, you can pass install with `pip install -e . -Cbuild-args=-j1` for example.

As for the implementation, it is ok to check if the component of the command starts with `--ninja-args`, then you need to parse the value passed to the option and add `-n` to the list. You can take some inspiration for how to do it from the code in this test https://github.com/mesonbuild/meson-python/blob/b3373346f4f60b9f0929d7a1042fa909badc6a48/tests/test_project.py#L49-L56

File: mesonpy/_editable.py
Comment (ogrisel) [reply mesonpy/_editable.py:336]: Sorry, I typed too quickly. I wanted to mention the editable modules of the tests.

File: .github/workflows/tests.yml
Comment (lesteve) [reply .github/workflows/tests.yml]: Yep I was about to do it :wink:, done in my last commit.

File: mesonpy/_editable.py
Comment (lesteve) [reply mesonpy/_editable.py]: I have changed the code to deal with existing and multiple `--ninja-args` in `self._build_cmd` but I haven't written a test for it yet.

File: mesonpy/_editable.py
Comment (lesteve) [reply mesonpy/_editable.py:345]: I have added a test to cover these lines in the main process, by adding a parametrized tests to `test_mesonpy_meta_finder` with `verbose=True`

> coverage (unless tricks are applied) does not cross process boundaries when collecting coverage data.

I have seen `parallel = True` in `.coveragerc` files to be able to cover lines in subprocesses, see [this](https://coverage.readthedocs.io/en/latest/cmd.html#combining-data-files-coverage-combine) for more details. Not sure if you would like to consider this option. We are doing this in scikit-learn and this seems to work fine but there may be some edge cases I am unaware of.

**Comment by dnicolodi** [comment]
Please don't merge the main branch into your branch the next time: it makes rebasing the commits harder without reason.

**Comment by dnicolodi** [comment]
I added the missing tests and fixed up the last issue and submitted as #594

**Pull request closed without merging**
